### DISCUSSION TITLE
How do a I restrict an RPC API (postgres function) to the service_role only

### DISCUSSION THREAD

**Discussion by jamesallchin** [supabase/supabase#3591] [Questions] [answered]
Hi, 

I have a postgres plpgsql function which gets exposed via Supabase Postgrest as an RPC API call. Very nice. 

However, in this particular use case I only want this API call to be callable by the special privileged service_key. 

In a non privilege scenario I can use `auth.uid()` to establish who the caller is. However, with the service key there doesn't seem to a way (at the DB level) to determine that the service account is being used. 

Is there a DB function to do this?

If I can check this, then I will be able to throw an exception when the caller is not using the service_key. 

Many thanks.

**Comment by jamesallchin** [comment]
Actually just answering my own question 😉  this might help me

https://github.com/supabase/supabase/discussions/2899

However, i think it would be a really useful feature to be able to access programatically in DB code is the calling user is the service_role. Does this exist?

**Comment by jamesallchin** [comment]
Also following up on :

https://github.com/supabase/supabase/discussions/2899

I had to carry out more revoke execute privileges to ensure that it could not be called by anon and authenticated users.

The complete steps were as follows.

```
REVOKE EXECUTE ON FUNCTION public.my_service_only_function(param1, param2) FROM anon;

REVOKE EXECUTE ON FUNCTION public.my_service_only_function(param1, param2) FROM authenticated;

REVOKE EXECUTE ON FUNCTION public.my_service_only_function(param1, param2) FROM postgres;

REVOKE EXECUTE ON FUNCTION public.my_service_only_function(param1, param2) FROM PUBLIC;

GRANT EXECUTE ON FUNCTION public.my_service_only_function(param1, param2) TO service_role;
```

**Comment by steve-chavez** [comment, marked as answer]
> If I can check this, then I will be able to throw an exception when the caller is not using the service_key.
> However, i think it would be a really useful feature to be able to access programatically in DB code is the calling user is the service_role.

Another option(if you don't want to do GRANTs) would be using `current_user` inside your function and throwing an exception if it doesn't match `service_role`.

```sql
create or replace function my_service_only_function() returns void as $$ begin

if current_user != 'service_role' then
    raise exception using message = 'denied';
end if;
-- ...
end $$ language plpgsql;
```

**Reply by jamesallchin** [reply]
Awesome thanks @steve-chavez - this is exactly what I was looking for 👍

**Comment by Odas0R** [comment]
Another solution:

```sql
create schema if not exists utils;

create or replace function utils.set_permission_on_execute (function_name text, role_name text)
returns void
language plpgsql
as $$
begin
  execute format('revoke execute on function %s from anon', function_name);
  execute format('revoke execute on function %s from authenticated', function_name);
  execute format('revoke execute on function %s from service_role', function_name);
  execute format('revoke execute on function %s from postgres', function_name);
  execute format('revoke execute on function %s from public', function_name);
  execute format('grant execute on function %s to %s', function_name, role_name);
end
$$;

-- set execution permissions to any function to any role
select utils.set_permission_on_execute('public.foo (uuid)', 'service_role');
select utils.set_permission_on_execute('public.bar (uuid)', 'service_role');
```
